fw/ss_skb.c

```diff
@@ -723,9 +723,20 @@ __skb_fragment(struct sk_buff *skb_head, struct sk_buff *skb, char *pspt,
 			goto done;
 		}
 
+		/*In case it was the last fragment of the current skb, we can
+		 * try to cut data from the the next skb.
+		 */
```

Comment:
I don't remember all the details, but it seemed to me that all requests to cut data from an skb were based on chunks, and a chunk didn't cross the boundaries of an skb or an skb paged fragment. So how come there's a request to cut data that crosses skb boundaries? Thanks.

Reply:
This is a debug output for this situation

```
[11596.494038] Start test: encoding.test_encoding.TestH2LargeResponseBodyDechunking2.test
[11596.572687] [tdb] Start Tempesta DB
[11596.705026] [tdb] Opened table /opt/tempesta/db/filter0.tdb: size=16777216 rec_size=20 base=00000000855157b4
[11596.799796] [tdb] Opened table /opt/tempesta/db/cache0.tdb: size=268435456 rec_size=0 base=00000000b2905084
[11596.807628] [tdb] Opened table /opt/tempesta/db/sessions0.tdb: size=16777216 rec_size=312 base=0000000052160c44
[11596.815957] [tdb] Opened table /opt/tempesta/db/client0.tdb: size=16777216 rec_size=624 base=000000004240708f
[11596.853060] [tempesta fw] Tempesta FW is ready
[11597.012412] j=16 offset=10 d_size=10
[11597.012413] j == si->nr_frags - 1
[11597.013706] [tempesta fw] SKB (ffff9921e1080000) DUMP: len=170 data_len=170 truesize=1146 users=1
[11597.014702] [tempesta fw]   head=ffff9921e1080100 data=ffff9921e1080240 tail=140 end=140
[11597.015644] [tempesta fw]   nr_frags=17 frag_list=0000000000000000 next=ffff9921e1080b80 prev=ffff9921a56e4700
[11597.016722] [tempesta fw]   head data (0):
[11597.017502] [tempesta fw]   frag  0 (addr=ffff99221af8ff06 pg_off=32518 size=10   pg_ref=2186):
[11597.018486]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.019432] [tempesta fw]   frag  1 (addr=ffff99221af8ff15 pg_off=32533 size=10   pg_ref=2186):
[11597.020445]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.021513] [tempesta fw]   frag  2 (addr=ffff99221af8ff24 pg_off=32548 size=10   pg_ref=2186):
[11597.022537]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.023523] [tempesta fw]   frag  3 (addr=ffff99221af8ff33 pg_off=32563 size=10   pg_ref=2186):
[11597.024640]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.025810] [tempesta fw]   frag  4 (addr=ffff99221af8ff42 pg_off=32578 size=10   pg_ref=2186):
[11597.026940]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.028019] [tempesta fw]   frag  5 (addr=ffff99221af8ff51 pg_off=32593 size=10   pg_ref=2186):
[11597.029252]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.030343] [tempesta fw]   frag  6 (addr=ffff99221af8ff60 pg_off=32608 size=10   pg_ref=2186):
[11597.031500]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.032572] [tempesta fw]   frag  7 (addr=ffff99221af8ff6f pg_off=32623 size=10   pg_ref=2186):
[11597.033820]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.034804] [tempesta fw]   frag  8 (addr=ffff99221af8ff7e pg_off=32638 size=10   pg_ref=2186):
[11597.035850]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.036876] [tempesta fw]   frag  9 (addr=ffff99221af8ff8d pg_off=32653 size=10   pg_ref=2186):
[11597.038100]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.039141] [tempesta fw]   frag 10 (addr=ffff99221af8ff9c pg_off=32668 size=10   pg_ref=2186):
[11597.040239]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.041390] [tempesta fw]   frag 11 (addr=ffff99221af8ffab pg_off=32683 size=10   pg_ref=2186):
[11597.042497]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.043554] [tempesta fw]   frag 12 (addr=ffff99221af8ffba pg_off=32698 size=10   pg_ref=2186):
[11597.044690]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.045870] [tempesta fw]   frag 13 (addr=ffff99221af8ffc9 pg_off=32713 size=10   pg_ref=2186):
[11597.046986]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.048045] [tempesta fw]   frag 14 (addr=ffff99221af8ffd8 pg_off=32728 size=10   pg_ref=2186):
[11597.049252]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.050319] [tempesta fw]   frag 15 (addr=ffff99221af8ffe7 pg_off=32743 size=10   pg_ref=2186):
[11597.051475]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.052545] [tempesta fw]   frag 16 (addr=ffff99221af8fff6 pg_off=32758 size=10   pg_ref=2186):
[11597.053856]     00000000: 71 77 65 72 74 79 75 69 6f 70                    qwertyuiop
[11597.054933] We have skb->next
[11597.055652] [tempesta fw] SKB (ffff9921e1080b80) DUMP: len=84772 data_len=84772 truesize=85668 users=1
[11597.056839] [tempesta fw]   head=ffff9921e1080c80 data=ffff9921e1080dc0 tail=140 end=140
[11597.058116] [tempesta fw]   nr_frags=6 frag_list=0000000000000000 next=ffff992213c1c380 prev=ffff9921e1080000
[11597.059371] [tempesta fw]   head data (0):
[11597.060185] [tempesta fw]   frag  0 (addr=ffff99221af90000 pg_off=0    size=124  pg_ref=2):
[11597.061586]     00000000: 0d 0a 61 0d 0a 71 77 65 72 74 79 75 69 6f 70 0d  ..a..qwertyuiop.
[11597.062843]     00000010: 0a 61 0d 0a 71 77 65 72 74 79 75 69 6f 70 0d 0a  .a..qwertyuiop..
[11597.064073]     00000020: 61 0d 0a 71 77 65 72 74 79 75 69 6f 70 0d 0a 61  a..qwertyuiop..a
[11597.065455]     00000030: 0d 0a 71 77 65 72 74 79 75 69 6f 70 0d 0a 61 0d  ..qwertyuiop..a.
[11597.066667]     00000040: 0a 71 77 65 72 74 79 75 69 6f 70 0d 0a 61 0d 0a  .qwertyuiop..a..
[11597.067885]     00000050: 71 77 65 72 74 79 75 69 6f 70 0d 0a 61 0d 0a 71  qwertyuiop..a..q
[11597.069210]     00000060: 77 65 72 74 79 75 69 6f 70 0d 0a 61 0d 0a 71 77  wertyuiop..a..qw
[11597.070430]     00000070: 65 72 74 79 75 69 6f 70 0d 0a 61 0d              ertyuiop..a.
[11597.071647] [tempesta fw] SKB (ffff9921e1080b80) DUMP: len=84767 data_len=84767 truesize=85668 users=1
[11597.073024] [tempesta fw]   head=ffff9921e1080c80 data=ffff9921e1080dc0 tail=140 end=140
[11597.074203] [tempesta fw]   nr_frags=6 frag_list=0000000000000000 next=ffff992213c1c380 prev=ffff9921e1080000
[11597.075598] [tempesta fw]   head data (0):
[11597.076444] [tempesta fw]   frag  0 (addr=ffff99221af90005 pg_off=5    size=119  pg_ref=2):
[11597.077830]     00000000: 71 77 65 72 74 79 75 69 6f 70 0d 0a 61 0d 0a 71  qwertyuiop..a..q
[11597.079079]     00000010: 77 65 72 74 79 75 69 6f 70 0d 0a 61 0d 0a 71 77  wertyuiop..a..qw
[11597.080282]     00000020: 65 72 74 79 75 69 6f 70 0d 0a 61 0d 0a 71 77 65  ertyuiop..a..qwe
[11597.081654]     00000030: 72 74 79 75 69 6f 70 0d 0a 61 0d 0a 71 77 65 72  rtyuiop..a..qwer
[11597.082836]     00000040: 74 79 75 69 6f 70 0d 0a 61 0d 0a 71 77 65 72 74  tyuiop..a..qwert
[11597.084041]     00000050: 79 75 69 6f 70 0d 0a 61 0d 0a 71 77 65 72 74 79  yuiop..a..qwerty
[11597.085351]     00000060: 75 69 6f 70 0d 0a 61 0d 0a 71 77 65 72 74 79 75  uiop..a..qwertyu
[11597.086579]     00000070: 69 6f 70 0d 0a 61 0d                             iop..a.
[11597.111880] j=0 offset=301392112 d_size=10
[11597.111899] j=13 offset=-35 d_size=10
[11597.381937] [tdb] Close table 'client0.tdb'
[11597.392159] [tdb] Close table 'sessions0.tdb'
[11597.459548] [tdb] Close table 'cache0.tdb'
[11597.495875] [tdb] Close table 'filter0.tdb'
[11597.497657] net_ratelimit: 14 callbacks suppressed
[11597.497658] [tempesta fw] modules are stopped
[11597.525133] [tempesta fw] exiting...
[11597.597180] [tdb] Shutdown Tempesta DB
[11597.696515] End test:   encoding.test_encoding.TestH2LargeResponseBodyDechunking2.test
```


**[11597.012412] j=16 offset=10 d_size=10**
In 16th fragment we have offset of 10 bytes form 10 bytes

**[11597.060185] [tempesta fw]   frag  0 (addr=ffff99221af90000 pg_off=0    size=124  pg_ref=2):
[11597.061586]     00000000: 0d 0a 61 0d 0a 71 77 65 72 74 79 75 69 6f 70 0d  ..a..qwertyuiop.
[11597.062843]     00000010: 0a 61 0d 0a 71 77 65 72 74 79 75 69 6f 70 0d 0a  .a..qwertyuiop..**
Here we have fragment 0 from next skb, where it is necessary to cut 5 first bytes.

Reply:
There is the code that makes this output

```
        if (unlikely(j >= 0)) {


		pr_notice("j=%i offset=%li d_size=%i", j, offset, d_size);

		if (len > 0) {
			if (!(ret = __new_pgfrag(skb_head, skb, len, j + 1, 1)))
				__it_next_data(skb, j + 1, it, fragn);
			goto done;
		}

		/*In case it was the last fragment of current skb, we can try to
		 * cut data from the next scb.
		 */
		if (j == si->nr_frags - 1) {
			pr_notice("j == si->nr_frags - 1\n");
			ss_skb_dump(skb);

			if (skb->next && (skb->next != skb_head)) {
				pr_notice("We have skb->next\n");
				ss_skb_dump1(skb->next);
				ret = __split_pgfrag_del_w_frag(skb_head,
				                                skb->next, 0,
				                                0, -len, it,
				                                fragn);
				ss_skb_dump1(skb->next);

			}
			else {
				T_WARN("Cannot delete bytes from skb\n");
				pr_notice("skb->next = null\n");\
				ret = -EFAULT;
			}
			goto done;
		}

		ret = __split_pgfrag_del_w_frag(skb_head, skb, j + 1, 0, -len,
		                                it, fragn);
		goto done;
	}
```
**ss_skb_dump1()** is the same as ss_skb_dump(), but outputs only the first fragment of skb.

Reply:
What I meant was that there's a few conventions:

- All stream data is accumulated into chunks, and no chunk can cross the boundaries of an skb or skb paged fragment;
- Request do cut data is based on the same chunks, and therefore can't cross these same boundaries.

That means that you shouln't (can't, really) get a request to cut data from an skb in Tempesta that would cross these  boundaries.

I am not sure what exactly is being done in the test.
Is it a request to remove 5 bytes, and the chunk where those 5 bytes are taken from somehow points to a wrong (previous) skb? If so, is that normal?
Or is it a request to remove data that somehow crosses the boundaries on an skb or skb paged fragment? It so, how can that happen?

Is there a clarified description of the case somewhere?
Thanks!